### PR TITLE
Implement support for calculating disk usage per tenant

### LINKED ISSUE #5961: Implement disk usage calculation for tenant level metrics
## Expected Behavior
Implement disk usage or simply `du` like functionality at per deployment level on per disk for more granular tenant level usage metrics. 

In future explore ways to enable AWS S3 CloudWatch like functionality https://docs.aws.amazon.com/AmazonS3/latest/dev/cloudwatch-monitoring.html to control metrics granularly.

## Current Behavior
The current behavior is at the disk level i.e `df -h` equivalent with no differentiation per tenant.

## Possible Solution
Do not have right now.

## Steps to Reproduce (for bugs)
This not a bug, it's a new feature 

## Context
Multi-tenant deployments

## Your Environment
master

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Implement support for calculating disk usage per tenant

## Motivation and Context
Fixes #5961

## How Has This Been Tested?
Manually using `pkg/madmin` API

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/posix.go [hunk 1]
```diff
@@ -285,6 +310,81 @@ func (s *posix) checkDiskFound() (err error) {
...
+	for {
+		select {
+		case <-doneCh:
+			return
+		case <-ticker.C:
+			usage, err := getDiskUsage(s.diskPath, buf, doneCh)
+			if err != nil {
+				continue
+			}
+			atomic.AddUint64(&s.usage, usage)
```
Comment (poornas) [inline cmd/posix.go]: diskUsage needs to be reset after each interval. Also, diskUsage seems to be calculated only after the 12 hour interval - needs to be calculated at server start as well.

**Review by poornas — CHANGES_REQUESTED** [review]

File: cmd/posix.go [hunk 2]
```diff
@@ -285,6 +310,81 @@ func (s *posix) checkDiskFound() (err error) {
 	return err
 }
 
+// stat information in parallel.
+func statInRoutine(dirPath string, entries []string, doneCh <-chan struct{}) <-chan os.FileInfo {
+	fInfoCh := make(chan os.FileInfo, 1)
+	go func() {
+		defer close(fInfoCh)
+		for _, entry := range entries {
+			fi, _ := os.Stat(pathJoin(dirPath, entry))
```
Comment (kannappanr) [inline cmd/posix.go]: Why are we ignoring error here?

Comment (harshavardhana) [reply cmd/posix.go]: Nothing to do with errors... at this point.

File: cmd/posix.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/posix.go]: done fixed.

File: cmd/fs-v1.go [hunk 3]
```diff
@@ -164,14 +171,43 @@ func (fs *FSObjects) Shutdown(ctx context.Context) error {
...
+	}
+	atomic.StoreUint64(&fs.usage, usage)
+
+	for {
+		select {
+		case <-doneCh:
+			return
+		case <-ticker.C:
+			usage, err := getDiskUsage(fs.fsPath, buf)
+			if err != nil {
```
Comment (kannappanr) [inline cmd/fs-v1.go]: Do we need to log this error? or will we be logging it in `getDiskUsage`?

Comment (harshavardhana) [reply cmd/fs-v1.go]: No, we don't need to log anything, there is nothing user can do. Other API calls will log and return appropriate errors.

File: cmd/posix-list-dir-nix.go [hunk 4]
```diff
@@ -168,3 +160,44 @@ func readDirN(dirPath string, count int) (entries []string, err error) {
 	}
 	return
 }
+
+// getDiskUsage returns du information for the posix path, recursively.
+func getDiskUsage(diskPath string, buf []byte) (uint64, error) {
```
Comment (kannappanr) [inline cmd/posix-list-dir-nix.go]: Just for clarification. Are we adding only the file sizes and not adding the directory metadata sizes?

Comment (harshavardhana) [reply cmd/posix-list-dir-nix.go]: Directories do not exist on object storage @kannappanr they are basically 0 bytes, so prefixes can be simply ignored. These metadata are negligible and serve no purpose.  We are only interested in what we created such as objects and other metadata JSON files etc.

File: cmd/posix-list-dir-others.go [hunk 5]
```diff
@@ -108,3 +108,44 @@ func readDirN(dirPath string, count int) (entries []string, err error) {
...
+
+// getDiskUsage returns du information for the posix path, recursively.
+func getDiskUsage(diskPath string, buf []byte) (uint64, error) {
+	d, err := os.Open(diskPath)
+	if err != nil {
+		return 0, err
+	}
+	defer d.Close()
+
+	maxEntries := 1000
```
Comment (kannappanr) [inline cmd/posix-list-dir-others.go]: Does this mean that we will support a maximum of 1000 buckets?

Comment (harshavardhana) [reply cmd/posix-list-dir-others.go]: No that is just for per Readdir() listing @kannappanr , the readdir proceeds to io.EOF

File: cmd/posix.go [hunk 6]
```diff
@@ -41,12 +42,24 @@ const (
...
 )
 
+const (
+	// readDirentBufSize for syscall.ReadDirent() to hold multiple
+	// directory entries in one buffer. golang source uses 4096 as
+	// buffer size whereas we want 64 times larger to save lots of
+	// entries to avoid multiple syscall.ReadDirent() call.
+	readDirentBufSize = 4096 * 64
+)
+
```
Comment (poornas) [inline cmd/posix.go]: since we are using non standard buffer size, is it meaningful to capture this as well along with usage?

Comment (harshavardhana) [reply cmd/posix.go]: why? it's not meaningful these are simply for dirents, dirents are not part of disk usage. This buffer contains filesystem direntinfo, it is part of filesystem tree, not the actual usage. We have always used a custom buffer to fetch larger entries in a single readdir operation. 

Disk usage here is simply calculating the total size of the files that are managed by Minio, nothing else.

File: cmd/posix-list-dir-nix.go [hunk 7]
```diff
@@ -168,3 +160,44 @@ func readDirN(dirPath string, count int) (entries []string, err error) {
...
+			break
+		}
+
+		var tmpEntries []string
+		if tmpEntries, err = parseDirents(diskPath, buf[:nbuf]); err != nil {
+			return 0, err
+		}
+
+		for _, entry := range tmpEntries {
+			// Look for only file entries.
```
Comment (poornas) [inline cmd/posix-list-dir-nix.go]: @harshavardhana , implementation here is ignoring all directories and not calculating du for contents of folders.. 
```
[images/ tbuck1/ nodedetails.yaml mybucket/ .minio.sys/ myb1/ userdata.yaml json.log bucket1/ tmp/]  in disk path /home/kris/export
➜  export du bucket1
112	bucket1
```

Comment (harshavardhana) [reply cmd/posix-list-dir-nix.go]: Yes, our `du` is not the same as filesystem `du` - our du is object storage `du` basically where there are only keys and size. Rest of the things are 0 bytes.

Comment (harshavardhana) [reply cmd/posix-list-dir-nix.go]: Oops you are talking about tree walk here, i see the problem. @poornas

File: pkg/madmin/examples/server-info.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
+// +build ignore
+
+/*
+ * Minio Cloud Storage, (C) 2016 Minio, Inc.
```
Comment (poornas) [inline pkg/madmin/examples/server-info.go]: 2016 -> 2018

**Review by poornas — APPROVED** [review]
LGTM.

**Review by kannappanr — APPROVED** [review]
Tested & LGTM

**Review by poornas — APPROVED** [review]

**Comment by poornas** [comment]
@harshavardhana ,browser tests failing on travis.

**Comment by harshavardhana** [comment]
> @harshavardhana ,browser tests failing on travis.

They are fixed @poornas

**Pull request merged**